core/src/test/java/io/undertow/server/ReadTimeoutTestCase.java

```diff
@@ -157,6 +174,62 @@ public long getContentLength() {
         }
     }
 
+    /**
+     * This verifies that read-timeout is cleaned-up after request is completely parsed. I.e. slow request processing
+     * happening after the request has been fully read should not trigger read-timeout.
+     */
+    @Test
+    public void testReadTimeoutWithSlowResponder() throws IOException {
+        DefaultServer.setRootHandler(new BlockingHandler((final HttpServerExchange exchange) -> {
+            FormParserFactory parserFactory = new FormParserFactory.Builder()
+                    .addParser(new FormEncodedDataDefinition())
+                    .build();
+
+            // Reads and parses the request.
+            try (FormDataParser formDataParser = parserFactory.createParser(exchange)) {
+                FormData formData = formDataParser.parseBlocking();
+                FormData.FormValue test = formData.getFirst("test");
+                Assert.assertNotNull(test);
+                Assert.assertEquals(FORM_PARAM_LENGTH, test.getValue().length());
+            }
+
+            // Write some response with delays, to breach read-timeout if it hasn't been cleaned-up.
+            Sender responseSender = exchange.getResponseSender();
+            for (int i = 0; i < 5; i++) {
+                Thread.sleep(200);
+                responseSender.send("*", new IoCallback() {
+                    @Override
+                    public void onComplete(HttpServerExchange exchange, Sender sender) {
+                        LOG.debug("Sent '*'");
+                    }
+
+                    @Override
+                    public void onException(HttpServerExchange exchange, Sender sender, IOException exception) {
+                        LOG.warn("Failed to write data to response channel.", exception);
+                    }
+                });
+            }
+        }));
+
+        try (TestHttpClient client = new TestHttpClient()) {
+            HttpPost post = new HttpPost(DefaultServer.getDefaultServerURL());
+
+            // Generate 2 KB form param value.
+            List<NameValuePair> nameValuePairs = new ArrayList<>();
+            StringBuilder sb = new StringBuilder(FORM_PARAM_LENGTH);
+            for (int i = 0; i < FORM_PARAM_LENGTH; i++) {
+                sb.append('a');
+            }
+            nameValuePairs.add(new BasicNameValuePair("test", sb.toString()));
+            post.setEntity(new UrlEncodedFormEntity(nameValuePairs));
+
+            // Request should succeed.
+            client.execute(post);
+        } catch (NoHttpResponseException e) {
+            Assert.fail("No request was received, this was presumably caused by read-timeout closing the connection.");
```

Comment:
Nitpicking, but the correct error message should be "No *response* was received from the server, ...", right?

Reply:
Oh right, that should say "response". Will update.